File: src/main/java/com/leijendary/spring/notificationtemplate/service/DeviceService.java
package com.leijendary.spring.notificationtemplate.service;

import com.leijendary.spring.notificationtemplate.data.DeviceData;
import com.leijendary.spring.notificationtemplate.factory.DeviceFactory;
import com.leijendary.spring.notificationtemplate.model.Device;
import com.leijendary.spring.notificationtemplate.repository.DeviceRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class DeviceService extends AbstractService {

    private final DeviceRepository deviceRepository;

    public Device save(final DeviceData deviceData) {
        final var deviceId = deviceData.getDeviceId();
        final var device = deviceRepository.findFirstByDeviceId(deviceId)
                .orElse(new Device());

        DeviceFactory.map(deviceData, device);

        return deviceRepository.save(device);
    }

    public void deleteByDeviceId(final String deviceId) {
        deviceRepository.deleteByDeviceId(deviceId);
    }
}

File: src/main/kotlin/com/leijendary/spring/template/notification/api/v1/service/DeviceService.kt
package com.leijendary.spring.template.notification.api.v1.service

import com.leijendary.spring.template.notification.api.v1.mapper.DeviceMapper
import com.leijendary.spring.template.notification.api.v1.model.DeviceDeregisterRequest
import com.leijendary.spring.template.notification.api.v1.model.DeviceRegisterRequest
import com.leijendary.spring.template.notification.client.NotificationClient
import com.leijendary.spring.template.notification.core.datasource.transactional
import com.leijendary.spring.template.notification.repository.DeviceRepository
import org.springframework.stereotype.Service
import java.util.*

@Service
class DeviceService(
    private val deviceRepository: DeviceRepository,
    private val notificationClient: NotificationClient
) {
    fun register(userId: UUID, request: DeviceRegisterRequest) {
        val platform = request.platform!!
        val token = request.token!!
        val device = deviceRepository
            .findFirstByToken(token)
            ?.apply { this.userId = userId }
            ?: DeviceMapper.INSTANCE.toEntity(userId, request)

        if (device.endpoint.isBlank()) {
            val endpoint = notificationClient.createEndpoint(platform, token) ?: return

            device.endpoint = endpoint
        }

        deviceRepository.save(device)
    }

    fun deregister(userId: UUID, request: DeviceDeregisterRequest) {
        val token = request.token!!
        val device = deviceRepository.findFirstByUserIdAndTokenOrThrow(userId, token)

        transactional {
            deviceRepository.delete(device)

            val endpoint = device.endpoint

            notificationClient.deleteEndpoint(endpoint)
        }
    }
}
